classes/03class/aws/README.md

```diff
@@ -0,0 +1,184 @@
+
+#### UserData
+
+The application deployment gets a little bit trickier as you will expect that everytime a new server comes up, to go and pickup your latest code to be deployed. So this is not a good candidate to be part of the instance image.
+
+The UserData is a script (bash or other languages) that you can make it run once your EC2 instance have the operational system ready to start.
+
+On this script you can for example, checkout a git repository, build your code and deploy to your running nginx webserver.
+
+
+## AWS Load Balancer
+
+Your webserver is now running on an ASG and being scaled up and down based on your computing resources needs or any other metrics that you chose. How would your users know to which instance they need to connect and how many of those are available at any point in time?
+
+That's how the load balancer works. Your clients will access through the load balancer and the LB will be in charge of balancing the requests across your instances.
+
+![ASG-LB](/classes/03class/aws/assets/ASG-LB.png)
+
+Load balancers can be of a few different types depending on what layer the routing decision should be:
+
+- Application Load Balancer
+  - application layer (HTTP/HTTPS)
+- Network Load Balancer
+  - transport layer (TCP/SSL)
+- Classic Load Balancer
+  - either application (HTTP/HTTPS) or transport (TCP/SSL) layer
+
+This material will be focused only on the Application Load Balancer which is used for most of the use cases.
+
+### Application Load Balancer (ALB)
+
+The service can be found at Services > Compute > EC2 > Load Balancing > Load Balancers > Create Load Balancer > ALB
+
+An ALB can serve as a single point of contact for clients and distribute traffic across multiple targets, such as EC2 Instances, in multiple Availability Zones. When creating an ALB you will have to chose from 2 different types and in which Availability Zones your LB will be available:
+
+- internet-facing: routes requests from public internet to targets
+- internal: routes requests only from clients to targets using private IP addresses
+
+Basically an ALB has two main components: listeners and target groups.
+
+```Load balancers are included on the AWS free-tier. However, please remember to delete any unused resources to avoid any costs```
+
+#### listeners
+
+A listener tries to match requests from clients based on, among many types, protocol/port and path-based routing. For example a request on port `443` can redirect the traffic to an internal server running on port `80` or a request to `/admin` sends the traffic to a specifc EC2 Instance.
+
+Each listener rule consists in a priority of actions and one or more conditions. When the conditions for a rule are met, the traffic is then forwarded to a target group.
```

Comment:
Each listener rule consists in a priority of actions and one or more conditions. When the conditions for a rule are met, the traffic is then forwarded to a target group and the remaining rules are ignored.